### ISSUE TITLE
[Impeller] Incorrect rendering of shadow with Impeller on iOS in stable.

### ISSUE THREAD

**Issue by vincent-hoodoo** [flutter/flutter#152357]
There is a bug on iOS with Impeller, where shadows will sometimes render as uniform rounded rectangles instead of the progressive shadow rendered with Skia.

I noticed that this bug was happening in our application only for some buttons, so I did some investigation and noticed that is seems to happen only for shadows of Widgets that are not aligned perfectly with the physical pixel "grid": in our app, it happens mostly for centered Widgets with subpixel dimensions.

I did a bit more investigation and managed to create a minimum reproducible example of a few lines only. See code below.

### Steps to reproduce

Create an empty iOS Flutter project:
```shell
flutter create test_project --platforms=ios --empty
cd test_project
dart pub get
```

Edit the `lib/main.dart` file to include the code detailed in the "Code sample" section below.

Run the app on the iPhone 15 simulator:
```shell
flutter run
```

### Expected results

Shadows should look correctly

### Actual results

Shadows appear as uniform grey rounded rectangles

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    final buttonStyle = ButtonStyle(
      elevation: WidgetStateProperty.all(4),
      padding: WidgetStateProperty.all(EdgeInsets.zero),
    );

    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              ElevatedButton(
                style: buttonStyle,
                onPressed: () {},
                child: const SizedBox(width: 142.5),
              ),
              ElevatedButton(
                style: buttonStyle,
                onPressed: () {},
                child: const SizedBox(width: 142.51),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```

</details>

##### First screenshot:
Difference between Impeller and Skia:   
![impeller_VS_skia](https://github.com/user-attachments/assets/d12f80b8-4c73-44bd-9f69-3945b56766ea)

##### Second screenshot:
This is a screenshot of the code sample running on the iPhone 15 simulator.
- the first button's shadow is incorrect
- the second button's shadow is correct

![Simulator Screenshot - iPhone 15 - 2024-08-01 at 14 56 21](https://github.com/user-attachments/assets/7305e4b2-d1b3-4362-90f7-2993c2e83c06)

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.3, on macOS 14.5 23F79 darwin-arm64, locale
    en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version
    34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.3)
[✓] VS Code (version 1.91.1)
[✓] VS Code (version 1.89.1)
[✓] Connected device (5 available)
[✓] Network resources

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
@vincent-hoodoo 
I tried below code sample with ElevatedButton but couldn't see the shadow or any rendering issue with Impeller:

<img width="317" alt="Screenshot 2024-07-26 at 3 55 59 PM" src="https://github.com/user-attachments/assets/6f299f40-52f4-491a-bfe2-a8f26bec9456">

```
body: Center(
        child: ElevatedButton(
          child: Text('Pay at Pump'),
          onPressed: () {},
        )
      )
```

If this doesn't represent your case, please provide complete runnable code sample.
See also if this is related to your case or not : https://github.com/flutter/flutter/issues/149757

**Comment by vincent-hoodoo** [comment]
@darshankawar I managed to create a minimum reproducible example, so I improved the description of the issue and added reproducing steps.

Please let me know if you can't reproduce it.

I'm surprised this hasn't been reported before, since Impeller has been around for a while now, and I guess that we're not the only one to come across it.

**Comment by darshankawar** [comment]
Thanks for providing reproducible code @vincent-hoodoo 
I tried it using latest master running on iphone 13 pro max simulator with and without impeller and get below results which seems to be similar as I couldn't find difference in rendering:

<img width="325" alt="Screenshot 2024-08-01 at 12 04 51 PM" src="https://github.com/user-attachments/assets/87975293-2d70-4f16-85eb-56a32a002a9b">

First screenshot is with Impeller, second is without it.

**Comment by vincent-hoodoo** [comment]
I just tried again, this time with flutter's later master, and I couldn't reproduce it.

So the bug is only on stable, but it must have been fixed since it's not on master anymore.

**Comment by darshankawar** [comment]
Thanks for the update. Yes, with the updated code, I was able to replicate the shadow behavior in latest stable but not in master, with and without Impeller.
 I'll keep the issue open and label for team's tracking and identifying which commit might have fixed it in master.

`Stable: 3.22.3`

**Comment by jonahwilliams** [comment]
I think this should be fixed in the latest stable (3.24), if not please reopen.

**Comment by darshankawar** [comment]
On 3.24.0, I get below running [this](https://github.com/flutter/flutter/issues/152357#issuecomment-2262269715) code sample on iphone 13 pro max simulator:

<img width="184" alt="Screenshot 2024-08-08 at 11 05 35 AM" src="https://github.com/user-attachments/assets/4b971638-8464-4354-9e83-5c38ce57ae0f">

@vincent-hoodoo please check the same at your end.
